devel/fortio/fortioLib/periodic.go

```diff
@@ -0,0 +1,244 @@
+	return &r.RunnerOptions // sort of returning this here
+}
+
+// Start starts the runner.
+func (r *periodicRunner) Run() {
+	useQPS := (r.QPS > 0)
+	var numCalls int64
+	if useQPS {
+		numCalls = int64(r.QPS * r.Duration.Seconds())
+		if numCalls < 2 {
+			log.Print("Increasing the number of calls to the minimum of 2 with 1 thread. total duration will increase")
+			numCalls = 2
+			r.NumThreads = 1
+		}
+		if int64(2*r.NumThreads) > numCalls {
+			r.NumThreads = int(numCalls / 2)
+			log.Printf("Lowering number of threads - total call %d -> lowering to %d threads", numCalls, r.NumThreads)
+		}
+		numCalls /= int64(r.NumThreads)
+		totalCalls := numCalls * int64(r.NumThreads)
+		fmt.Printf("Starting at %g qps with %d thread(s) [gomax %d] for %v : %d calls each (total %d)\n",
+			r.QPS, r.NumThreads, runtime.GOMAXPROCS(0), r.Duration, numCalls, totalCalls)
+	} else {
+		fmt.Printf("Starting at max qps with %d thread(s) [gomax %d] for %v\n",
+			r.NumThreads, runtime.GOMAXPROCS(0), r.Duration)
+	}
+	start := time.Now()
+	// Histogram  and stats for Function duration - millisecond precision
+	functionDuration := NewHistogram(0, r.Resolution)
+	// Histogram and stats for Sleep time (negative offset to capture <0 sleep in their own bucket):
+	sleepTime := NewHistogram(-0.001, 0.001)
+	if r.NumThreads <= 1 {
+		if r.Verbose > 0 {
+			log.Printf("Running single threaded")
+		}
+		runOne(0, functionDuration, sleepTime, numCalls, start, r)
+	} else {
+		var wg sync.WaitGroup
+		var fDs []*Histogram
+		var sDs []*Histogram
+		for t := 0; t < r.NumThreads; t++ {
+			durP := functionDuration.Clone()
+			sleepP := sleepTime.Clone()
+			fDs = append(fDs, durP)
+			sDs = append(sDs, sleepP)
+			wg.Add(1)
+			go func(t int, durP *Histogram, sleepP *Histogram) {
+				defer wg.Done()
+				runOne(t, durP, sleepP, numCalls, start, r)
+			}(t, durP, sleepP)
+		}
+		wg.Wait()
+		for t := 0; t < r.NumThreads; t++ {
+			functionDuration.Transfer(fDs[t])
+			sleepTime.Transfer(sDs[t])
+		}
+	}
+	elapsed := time.Since(start)
+	actualQPS := float64(functionDuration.Count) / elapsed.Seconds()
+	fmt.Printf("Ended after %v : %d calls. qps=%.5g\n", elapsed, functionDuration.Count, actualQPS)
+	if useQPS {
+		percentNegative := 100. * float64(sleepTime.hdata[0]) / float64(sleepTime.Count)
+		if percentNegative > 3 {
+			sleepTime.FPrint(os.Stdout, "Aggregated Sleep Time", 50)
+			fmt.Printf("WARNING %.2f%% of sleep were falling behind\n", percentNegative)
+		} else {
+			if Verbose > 0 {
+				sleepTime.FPrint(os.Stdout, "Aggregated Sleep Time", 50)
+			} else {
+				sleepTime.Counter.FPrint(os.Stdout, "Sleep times")
+			}
+		}
+	}
+	functionDuration.FPrint(os.Stdout, "Aggregated Function Time", r.Percentiles[0])
+	for _, p := range r.Percentiles[1:] {
+		fmt.Printf("# target %g%% %.6g\n", p, functionDuration.CalcPercentile(p))
+	}
+}
+
+func runOne(t int, cF *Histogram, cS *Histogram, numCalls int64, start time.Time, r *periodicRunner) {
```

Comment:
There is a strong preference for short variable names in golang. However, function parameters probably ought to be a bit more verbose since they serve as documentation for the user. In this particular case I had no idea `t`, `cF`, and `cS` were until I read the caller function. Something like `streamID`, `freq` and `sleep` respectively might be more self describing.